### PR TITLE
Fixes markdown parse error #127

### PR DESCRIPTION
**Pull request by alexkart**
Try to send a message to telegram without parsemode markdown if it can't be parsed correctly.

### CODE REVIEW

**Comment by umputun** [comment]
Thanks for the PR. To me, it sounds like a good candidate for a dedicated function, like `sendMdWithFailback` or something similar. It would also be nice to have some log message in such a case. Additionally, having a test for this change will help.

btw, a similar code already exists in tg-spam, doing almost the same thing https://github.com/umputun/tg-spam/blob/8f6edd06cc4fcea1359a70fe95f48adb299d065a/app/events/events.go#L71-L71

**Comment by alexkart** [comment]
Okay, I'll extract a function, but I may still need some help since I'm just a beginner in Go.

**Comment by umputun** [comment]
> but I may still need some help since I'm just a beginner in Go

Sure, feel free to ping @paskal if you need something Telegram-specific. He contributed a lot of Telegram functionality to this bot.

**Comment by alexkart** [comment]
Extracted the `sendMdWithFallback` function and added tests and logging.

**Comment by alexkart** [comment]
Can someone review this PR? @umputun @paskal

File: app/events/telegram.go [hunk 1]
```diff
@@ -276,6 +272,34 @@ func (l *TelegramListener) sendBotResponse(resp bot.Response, chatID int64) erro
 	return nil
 }
 
+// sendMdWithFallback sends message with markdown mode and fallback to plain text
+func (l *TelegramListener) sendMdWithFallback(resp bot.Response, chatID int64) (tbapi.Message, error) {
+	log.Printf("[DEBUG] sending message to telegram")
+	tbMsg := tbapi.NewMessage(chatID, resp.Text)
+	tbMsg.ParseMode = tbapi.ModeMarkdown
+	if resp.ParseMode != "" {
+		tbMsg.ParseMode = resp.ParseMode
```
Comment (paskal) [inline app/events/telegram.go:281]: The unfilled `ParseMode` switches to [legacy markdown](https://core.telegram.org/bots/api#markdown-style). Please see [documentation](https://core.telegram.org/bots/api#formatting-options) for details. So, the problem you are trying to solve will still need to be solved. In that case, you should apply [EscapeTelegramText](https://github.com/go-pkgz/notify/blob/master/telegram.go#L150C6-L156).

Another solution would be escaping a Telegram message according to prohibited symbols in the documentation, but there is no ready-made function for that, and if you add it, it's better to add it to the go-pkgz/notify telegram.go file.

Comment (umputun) [reply app/events/telegram.go:281]: I have not found a 100% working way to escape. I used the same function, even extended with more use cases, and still get invalid markdown format errors of all kinds, from time to time. I believe the proposed solution is supposed to do a fallback to plain text. I didn't read the docs, but from my experience, setting ParseMode to an empty string made Telegram ignore the markdown. I'm not sure if this is plain text mode or some limited markdown, but the result looks like plain text to me. This is one of the examples of a message that failed after the escape and was sent with an empty parse mode.

image

similar one with a valid tg markdown looks like this

image

Comment (alexkart) [reply app/events/telegram.go:281]: Yes, I also did some testing in Telegram chat, it also looked like plain text, I didn't notice any formatting.

Comment (paskal) [reply app/events/telegram.go:281]: I think you two are right, and I am wrong. It's not written explicitly but it seems like the default way is indeed undocumented PlainText and not legacy Markdown.

**Review by paskal — APPROVED** [review]

File: app/events/telegram.go [hunk 2]
```diff
@@ -276,6 +272,34 @@ func (l *TelegramListener) sendBotResponse(resp bot.Response, chatID int64) erro
...
+	}
+	tbMsg.DisableWebPagePreview = !resp.Preview
+	tbMsg.ReplyToMessageID = resp.ReplyTo
+	res, err := l.TbAPI.Send(tbMsg)
+
+	if err != nil {
+		// If it can't parse entities, try to send message as plain text
+		if tbMsg.ParseMode == tbapi.ModeMarkdown && strings.Contains(err.Error(), "Bad Request: can't parse entities:") {
+			log.Printf("[WARN] failed to send message as markdown, %v", err)
```
[the remark this answers is not in the document]
Comment (alexkart) [reply app/events/telegram.go]: Done.

**Pull request merged**